### PR TITLE
Generate Screenshots in CI

### PR DESCRIPTION
**Pull request by jkmassel**
**To Test:**

**Part 1:**
- See that the Checks list shows that all jobs are passing – that means the screenshots were generated successfully.

**Part 2:**
- Create a new branch from this one, and create a PR for it. 
- Note that the PR by default doesn't create screenshots.
- Add the "Needs Screenshots" label to the PR – that'll cause the process to start.
- Verify that it succeeds (it usually takes about an hour, give or take)

**Cleanup:**
- Close down your test PR and delete the test branch.

**Note:**
This functionality can still be a little flaky at times – based on the logs and automated screenshots captured during failures, this seems to be an issue on Xcode's part – it sometimes can't properly launch and communicate with the app in the simulator. Because it's so intermittent, and there's not much we can do about it, IMHO the functionality offered by this PR is good enough that we ought to ship it – if we want to re-run the screenshot generation code, we can do so by removing and re-adding the label, and it's still often faster than running this all locally.

### CODE REVIEW

File: .github/workflows/screenshots.yml [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+      id: restore-ruby-dependency-cache
+      uses: actions/cache@v1
+      with:
+        path: vendor/bundle
+        key: ${{ runner.os }}-gems-${{ hashFiles('**/Gemfile.lock') }}
+    
+    - name: Install Ruby Dependencies
+      run: |
+        bundle config path vendor/bundle
+        bundle install --jobs 4 --retry 3
```
Comment (mokagio) [inline .github/workflows/screenshots.yml]: I was gonna suggest to use an `if` to check for `cache-hit` to skip running `bundle install` if the cache in the step above was hit (docs [here](https://github.com/actions/cache/tree/e8230b28a9d4dd887adc2d080b81482b2d47b185#skipping-steps-based-on-cache-hit), but I can see that both the [Ruby example](https://github.com/actions/cache/blob/e8230b28a9d4dd887adc2d080b81482b2d47b185/examples.md#ruby---bundler) and the [Actions docs](https://help.github.com/en/actions/configuring-and-managing-workflows/caching-dependencies-to-speed-up-workflows#example-using-the-cache-action) don't do it. Do you know why?

File: .github/workflows/screenshots.yml [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+        path: fastlane/DerivedData/Build/Products/Debug-iphonesimulator/WooCommerce.app
+
+    - name: Download Screenshot Runner
+      uses: actions/download-artifact@v1
+      with:
+        name: screenshot-runner
+        path: fastlane/DerivedData/Build/Products/Debug-iphonesimulator/WooCommerceScreenshots-Runner.app
+
+    - name: Start Mocks
+      run: rake mocks &
```
Comment (mokagio) [inline .github/workflows/screenshots.yml]: Clever!

File: .github/workflows/screenshots.yml [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+
+    - name: Generate Screenshots
+      run: |
+        bundle exec fastlane take_screenshots languages:${{ matrix.language }} mode:${{ matrix.mode }}
+
+    - name: Move Test Result Bundle to Logs
+      if: always()
```
Comment (mokagio) [inline .github/workflows/screenshots.yml]: Is there a particular use case for `if: always()`?

File: .github/workflows/screenshots.yml [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+      id: restore-ruby-dependency-cache
+      uses: actions/cache@v1
+      with:
+        path: vendor/bundle
+        key: ${{ runner.os }}-gems-${{ hashFiles('**/Gemfile.lock') }}
+    
+    - name: Install Ruby Dependencies
+      run: |
+        bundle config path vendor/bundle
+        bundle install --jobs 4 --retry 3
```
Comment (mokagio) [inline .github/workflows/screenshots.yml]: This matches the example in the

File: .github/workflows/screenshots.yml [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+env:
+  S3_BUCKET: ${{ secrets.S3_BUCKET }}
+  AWS_ACCESS_KEY_ID: ${{ secrets.AWS_ACCESS_KEY_ID }}
+  AWS_SECRET_ACCESS_KEY: ${{ secrets.AWS_SECRET_ACCESS_KEY }}
+  CONFIGURE_ENCRYPTION_KEY: ${{ secrets.CONFIGURE_ENCRYPTION_KEY }}
+
+jobs:
+  build:
+    name: Build Application
+    if: contains(github.event.pull_request.labels.*.name, 'Needs Screenshots')
```
Comment (mokagio) [inline .github/workflows/screenshots.yml]: This `if` is super cool.

File: .github/workflows/screenshots.yml (hunk 1 above)
Comment (jkmassel) [reply .github/workflows/screenshots.yml]: I'm not 100% sure why, but I found that just restoring from cache would sometimes have weird failures (particularly on our other dependencies restoration). 

Allowing `bundle install` to run after restoring the cache only takes ~1 second, so whatever bookkeeping it's doing, it makes everything work :)

File: .github/workflows/screenshots.yml (hunk 3 above)
Comment (jkmassel) [reply .github/workflows/screenshots.yml]: Yeah! That syntax causes the step to run even if the preceding steps fail – it's good for getting logs off a CI machine that's failed a step!

File: .github/workflows/screenshots.yml (hunk 1 above)
Comment (mokagio) [reply .github/workflows/screenshots.yml]: Great to know! I'll update my setup as well then 👌

**Comment by rachelmcr** [comment]
@jkmassel I tried this out in https://github.com/woocommerce/woocommerce-ios/pull/1901 and my first attempt resulted in [this failure](https://github.com/woocommerce/woocommerce-ios/pull/1901/checks?check_run_id=469619283) caused by a timeout:

> xcodebuild -showBuildSettings timed out after 4 retries with a base timeout of 3. You can override the base timeout value with the environment variable FASTLANE_XCODEBUILD_SETTINGS_TIMEOUT, and the number of retries with the environment variable FASTLANE_XCODEBUILD_SETTINGS_RETRIES 

Have you seen that before? I thought about just re-trying it right away but I wanted to preserve this in case there are timeout issues we need to resolve to make this stable.

(I also initially pushed that branch without adding another commit to it, and I think that caused the failed check on this PR. Sorry about that!)

**Comment by jkmassel** [comment]
No worries about the push!!

Yeah that's Xcode being annoying. I'd say just re-run it – there's really nothing we can do about that case (and I haven't seen it before).

**Comment by jkmassel** [comment]
Screen Shot 2020-02-27 at 8 11 44 AM

@rachelmcr – this was some silliness with CircleCI – I re-ran it, and it passed :)

**Pull request merged**